feat(noSecrets): refine the entropy computation to avoid some false positives

SaadBazaz opened the pull request:
<!--
	Thanks for submitting a Pull Request! We appreciate you spending the time to work on these changes.
	Please provide enough information so that others can review your PR.
	Once created, your PR will be automatically labeled according to changed files.
	Learn more about contributing: https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md
-->

Tackles #4113 and #3861

Further improves #3823 

### Research
Ever since the creation of this biome plugin originally, I've found various other secret scanning solutions:

1. The OG, gitleaks (Go): https://github.com/gitleaks/gitleaks/
2. Trufflehog (Go): https://github.com/trufflesecurity/trufflehog
3. Sensleak (Rust): https://github.com/crates-pro/sensleak-rs

I've been wondering; Can we consume either one of the above to use in our usecase? Do we really need to add another to the ecosystem?
This I'd like to ask the authors here.

In the meanwhile...
### Fixes
- [x] Fix original eslint attribution
- [x] Add better tests for generic use cases

### New features planned

**Legend:**
FP = Feature-Parity with eslint no-secrets (features listed [here](https://github.com/nickdeis/eslint-plugin-no-secrets?tab=readme-ov-file#4-options))

- [ ] Add config:
  - [x] Control the entropy (FP)
  - [ ] add excludeRegexes
  - [ ] add additionalRegexes (FP)
  - [ ] checkComments
  - [ ] ignoreIdentifiers (FP)
- [ ] Add better entropy mechanism
- [ ] Add more stopwords and heuristics
  - [x] Contains space
  - [ ] Startswith check
- [ ] Check for secrets in comments (FP)
- [ ] Pass all current tests (FP)

I would love to hear strategies on how to achieve this.

dyc3 commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -36,8 +36,8 @@ declare_lint_rule! {
         language: "js",
-        recommended: false,
-        sources: &[RuleSource::Eslint("no-secrets/no-secrets")],
+        recommended: true,
```
```suggestion
        recommended: false,
```

SaadBazaz commented:
@minht11 @Conaclos @dyc3  - Would appreciate if you can read the PR description and drop your thoughts.

dyc3 commented:
I'll defer to the other maintainers for their opinions on adding a dependency for this.

For the proposed options: We generally try to avoid adding options for the sake of adding options. We need an appropriate amount of demand (from users) and justification to determine the correct granularity and scope of those options.

>  Can we consume either one of the above to use in our usecase? Do we really need to add another to the ecosystem?

Personally, I think a dedicated tool is probably going to have better heuristics than we will, at least in the short term. Plus, we don't take commit history into account like some of those tools do (and we shouldn't, that's not what biome is for).

I think this rule will at least be good for picking out the most egregious cases, with the added benefit that adding the rule to their existing config is easier than adding a new tool to their chain. We could also consider pointing users to a dedicated tool in the diagnostic message and rule documentation. Gitleaks seems to be the most mature.

minht11 commented:
To add to what @dyc3 said, in my opinion Biome likely isn't a right tool to fully replace full repo secret scanning. Biome does not support all the languages which those tools do and need to be a proper security solution, we have stricter perf constraints and so on.

Having this rule in Biome does help development velocity, no need to wait for CI, and majority of users who wouldn't think about secrets get a warning for common cases, though we definitely should document about full solutions in case users need those.

As for integrations listed:
Sensleak - does not have rust crate, only CLI.
Otherss are GO solutions. Not sure how portable they are, can you call GO lib from rust reasonably? How would the performance be?

dyc3 commented:
> can you call GO lib from rust reasonably?

No. Turbopack had a whole zig helper library to interop between rust and golang while they were porting the project to rust. Definitely not worth the complexity.

SaadBazaz commented:
> For the proposed options: We generally try to avoid adding options for the sake of adding options. We need an appropriate amount of demand (from users) and justification to determine the correct granularity and scope of those options.

I agree!

There are some features which would make us feature-parity with eslint no-secrets, I've updated the description with a legend to reflect that.

> We could also consider pointing users to a dedicated tool in the diagnostic message and rule documentation. Gitleaks seems to be the most mature.
> ... though we definitely should document about full solutions in case users need those.

Definitely going to do this!

> Sensleak - does not have rust crate, only CLI.

I can port some code then 🤠 Or maybe ask their team to make a crate/API for reusability.

> No. Turbopack had a whole zig helper library to interop between rust and golang while they were porting the project to rust. Definitely not worth the complexity.

Yikes.

ematipico commented:
We usually add options only when required or requested, and when there are valid use cases to cover.

Not sure there's enough value to add yet another dependency for this rule. Sure, it's an important rule, but as the others said, we don't cover all languages. Our documentation could actually propose alternatives to the users.

Conaclos commented:
I totally agree with others. The rule should find the most obvious secret leaks and avoid false positives because users might get annoyed and turn off the rule (that defeat its purpose).

We could add a disclaimer in the rule description saying that and point to relevant tools (as the one you cited) for advanced secret leak detections.

SaadBazaz commented:
> I totally agree with others. The rule should find the most obvious secret leaks and avoid false positives because users might get annoyed and turn off the rule (that defeat its purpose).
> 
> We could add a disclaimer in the rule description saying that and point to relevant tools (as the one you cited) for advanced secret leak detections.

In that case, (which I agree with), we should just do two tasks for now:
1. Add the containsSpace heuristic which will drastically reduce false positives
2. Update the docs and error message to suggest relevant tools

However I do suggest **3) adding the option to control entropy** as it's not only easy to add, but also useful for us to learn what people are usually comfortable with so we can improve the default option.

What do you think?

Conaclos commented:
> However I do suggest 3) adding the option to control entropy as it's not only easy to add, but also useful for us to learn what people are usually comfortable with so we can improve the default option.

This looks fair enough to me.

SaadBazaz commented:
@Conaclos - Noob question:
```
#[derive(Clone, Debug, Default, Deserialize, Deserializable, Eq, PartialEq, Serialize)]
#[cfg_attr(feature = "schema", derive(schemars::JsonSchema))]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct NoSecretsOptions {
    /// Set entropy threshold (default is 4.5).
    entropy_threshold: f64,
}
```

How can I get f32 or f64 options (i.e. any floating point option) in Options?

Conaclos commented:
> How can I get f32 or f64 options (i.e. any floating point option) in Options?

`ctx.options().entropy_threshold` in the `run`/`diagnostic` methods.

SaadBazaz commented:
@Conaclos - I'm getting the following error:
```bash
error[E0277]: the trait bound `f64: std::cmp::Eq` is not satisfied
   --> crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:135:5
    |
130 | #[derive(Clone, Debug, Default, Deserialize, Deserializable, Eq, PartialEq, Serialize)]
    |                                                              -- in this derive macro expansion
...
135 |     entropy_threshold: f64, // @TODO: Doesn't work currently.
    |     ^^^^^^^^^^^^^^^^^^^^^^ the trait `std::cmp::Eq` is not implemented for `f64`
    |
    = help: the following other types implement trait `std::cmp::Eq`:
              i128
              i16
              i32
              i64
              i8
              isize
              u128
              u16
            and 4 others
note: required by a bound in `AssertParamIsEq`
```

Reading the docs:
> The compiler should warn you that MyRuleOptions does not implement some required types. We currently require implementing serde's traits Deserialize/Serialize.

https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md#:~:text=The%20compiler%20should%20warn%20you%20that%20MyRuleOptions%20does%20not%20implement%20some%20required%20types.%20We%20currently%20require%20implementing%20serde%27s%20traits%20Deserialize/Serialize.

It seems like f64 isn't supported in Options, am I missing out on something?

Conaclos commented:
@SaadBazaz

`f64` and `f32` doesn't implement `Eq`/`PartialEq` because directly comparing floats is an error.
I am unsure why we need the options to be `Eq`/`PartialEq`.
I have to investigate if we can remove it.

For now, I see two possible approaches:

1. Implement ourselves `PartialEq`.
   Usually we compare floats by taking the absolute value of their subtraction and comparing it as EPSILON.
   ```rs
   impl Eq for MyOptions {}
   impl PartialEq for MyOptions {
        fn eq(&self, other: &Self) -> bool {
            (self.entropy_threshold - other.entropy_threshold).abs() < f64:EPSILON
        }
   }
   ```
2. Use an integer instead of float for the threshold.
   For example, we could use a `u16`.
   We could either divide the provided threshold by 10 (or even 100 if we need more precision: not sure if we need it) before comparing it against the computed threshold.
  Or we could make the reverse: multiply by 10 (or 100) the computed threshold and truncate it before comparing it against the provided threshold.

Personnaly I could choose (2).

SaadBazaz commented:
> Personnaly I could choose (2).

Yeah I thought about (2) too, it'll allow us to have an abstraction for later on when/if we change the underlying entropy function. I'll go for that.

SaadBazaz commented:
- [x] Updated entropy function to add a few more metrics which I thought might be useful (i.e. checking consecutive upper/lower cases, checking presence of numbers and symbols)
- [x] Had to comment out some tests :( I would love to cater them all but too many hyperparameters to balance. Need to go deeper into entropy or have better regexes.
- [x] Added entropy_threshold as Option.
- [x] Updated docs
- [x] Updated error message
- [x] Updated inspiration to original plugin

SaadBazaz commented:
@Conaclos @dyc3 Would love a review on the work so far.

dyc3 commented on crates/biome_js_analyze/tests/specs/nursery/noSecrets/valid.js:
```diff
@@ -4,3 +4,39 @@ const count = 10;
+// TODO: Remove these false positives, they unfortunately hurt the user experience.
+// const NAMESPACE_CLASSNAME = 'Validation.JSONSchemaValidationUtilsImplFactory';
+// const BASE64_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";
+// const webpackFriendlyConsole = require('./config/webpack/webpackFriendlyConsole');
```
We could easily exclude strings inside `require()` and `import()` calls

SaadBazaz replied:
Would that be by looking at the sibling / previous node in the AST?

dyc3 replied:
The parent, but yeah.

SaadBazaz replied:
We can do that in V3 alongside implementing comments, what say? Because we haven't involved the AST as of yet.

dyc3 reviewed (commented):
Just a quick review.

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -10,14 +10,28 @@ use regex::Regex;
+    /// sensitive data exposure.
     ///
-    /// While this rule is helpful, it's not infallible. Always review your code carefully and consider implementing additional security measures like automated secret scanning in your CI/CD and git pipeline, such as GitGuardian or GitHub protections.
+    /// ## Recommendations
```
We should add a blank line before and after every heading.
```suggestion
    ///
    /// ## Recommendations
    ///
```

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -155,119 +198,272 @@ struct SensitivePattern {
+    base_threshold + entropy * scaling_adjustment
+}
+
+fn split_into_tokens(value: &str) -> Vec<&str> {
+    let delimiters = [' ', '\t', '\n', '.', ',', ';', ':', '/', '-', '_', '@'];
+    let mut tokens = vec![value];
+
+    for &delimiter in delimiters.iter() {
+        tokens = tokens
+            .into_iter()
+            .flat_map(|token| token.split(delimiter).filter(|&s| !s.is_empty()))
+            .collect();
+    }
+
+    tokens
```
Have you tried [str::split](https://doc.rust-lang.org/std/primitive.str.html#method.split)?

SaadBazaz replied:
Works. Went for that instead, thanks for catching.

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -155,119 +198,272 @@ struct SensitivePattern {
+    base_threshold: f64,
+    scaling_factor: f64,
+) -> f64 {
+    // We will apply a logarithmic dampening to prevent excessive scaling for long tokens
```
Have you some links to add in comment about this?

SaadBazaz replied:
Added some [here](https://github.com/biomejs/biome/pull/4118/files#diff-f690e55d08175a773832c214f5d8e28473a929beca5a75a7dc9b41143fc48796R481). Also added references in [Shannon Entropy](https://github.com/biomejs/biome/pull/4118/files#diff-f690e55d08175a773832c214f5d8e28473a929beca5a75a7dc9b41143fc48796R390).

minht11 commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -10,14 +10,28 @@ use regex::Regex;
     ///
-    /// While this rule is helpful, it's not infallible. Always review your code carefully and consider implementing additional security measures like automated secret scanning in your CI/CD and git pipeline, such as GitGuardian or GitHub protections.
+    /// ## Recommendations
+    /// While this rule is beneficial for catching the most egregious cases,
```
Could we simplify language a bit as not a native speaker words like egregious and infallible are bit hard to parse.
Maybe something like this:

"While this rule helps with most common cases it is not intended to handle all of them."

SaadBazaz replied:
Done! Good point!

minht11 commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
````diff
@@ -32,12 +46,17 @@ declare_lint_rule! {
     /// ```
+    ///
+    /// ## Disclaimer
+    /// This rule is intended to catch obvious secret leaks, but for more robust detection
````
This section repeats what other above just said. Maybe we could leave short sentence above and move exact tool recommendations and explanation here?

minht11 commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -10,14 +10,28 @@ use regex::Regex;
     ///
     /// This rule checks for high-entropy strings and matches common patterns
-    /// for secrets, such as AWS keys, Slack tokens, and private keys.
+    /// for secrets, including AWS keys, Slack tokens, and private keys.
```
Should we list explicitly what secrets we try to catch? Now the only way to know would be looking into source code or blindly guessing.

SaadBazaz replied:
Good point!

SaadBazaz replied:
Added.

SaadBazaz commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -92,16 +121,37 @@ impl Rule for NoSecrets {
                 "\n2. If needed, use environment variables or a secure secret management system to store sensitive data."
-                "\n3. If this is a false positive, consider adding an inline disable comment."
+                "\n3. If this is a false positive, consider adding an inline disable comment, or tweak the entropy threshold. Learn more: https://biomejs.dev/linter/rules/no-secrets/#options"
+                "\nThis rule only catches very basic vulnerabilities. For more robust, proper solutions, we recommend heading over to https://biomejs.dev/linter/rules/no-secrets/#recommendations"
```
Added!

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -155,119 +223,284 @@ struct SensitivePattern {
+}
+
+fn split_into_tokens(value: &str) -> Vec<&str> {
+    let delimiters = [' ', '\t', '\n', '.', ',', ';', ':', '/', '-', '_', '@'];
+    value
+        .split(|c| delimiters.contains(&c))
```
You can directly pass the list of separators.
```suggestion
    value
        .split([' ', '\t', '\n', '.', ',', ';', ':', '/', '-', '_', '@'])
```

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -155,119 +223,284 @@ struct SensitivePattern {
+}
+
+fn detect_secret(data: &str, entropy_threshold: &u16) -> Option<&'static str> {
+    let tokens: Vec<&str> = split_into_tokens(data);
```
We could directly call split/filter here and avoid allocating a `Vec` by iterating over the iterator.
Another possibility is to keep `split_into_tokens` and returning `impl Iterator<Item = str>` instead of `Vec<&str>`.
```suggestion
    let tokens = data.split(...).filter(...);
```

SaadBazaz replied:
Keeping the former option because the other refactors have reduced the code to 3 lines 😂 👌

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -155,119 +223,284 @@ struct SensitivePattern {
+    }
+
+    // Letter classification and case switching
+    for (i, c) in data.chars().enumerate() {
+        if c.is_ascii_alphabetic() {
+            letter_count += 1;
+            if c.is_uppercase() {
+                uppercase_count += 1;
+                if i > 0 && !previous_char_was_upper {
+                    case_switches += 1;
+                }
+                previous_char_was_upper = true;
+            } else {
+                lowercase_count += 1;
+                if i > 0 && previous_char_was_upper {
+                    case_switches += 1;
+                }
+                previous_char_was_upper = false;
+            }
+        } else if c.is_ascii_digit() {
+            digit_count += 1;
+        } else if !c.is_whitespace() {
+            symbol_count += 1;
```
Since we check only ASCII stuff, we could directly iterate over bytes. This is more efficient because this avoids decoding UTF-8 codepooints.

```suggestion
    for (i, c) in data.bytes().enumerate() {
        if c.is_ascii_alphabetic() {
            letter_count += 1;
            if c.is_ascii_uppercase() {
                uppercase_count += 1;
                if i > 0 && !previous_char_was_upper {
                    case_switches += 1;
                }
                previous_char_was_upper = true;
            } else {
                lowercase_count += 1;
                if i > 0 && previous_char_was_upper {
                    case_switches += 1;
                }
                previous_char_was_upper = false;
            }
        } else if c.is_ascii_digit() {
            digit_count += 1;
        } else if !c.is_ascii_whitespace() {
            symbol_count += 1;
```

Conaclos commented on crates/biome_js_analyze/src/lint/nursery/no_secrets.rs:
```diff
@@ -155,119 +223,284 @@ struct SensitivePattern {
-fn calculate_shannon_entropy(data: &str) -> f64 {
+/*
+Uses Shannon Entropy as a base algorithm, then adds "boosts" for special patterns/occurrences.
+For example, Continuous mixed cases (lIkE tHiS) are more likely to contribute to a higher score than single cases.
+Symbols also contribute highly to secrets.
```
Have you based this on existing works? This could be worth adding references.

SaadBazaz replied:
Not really, I did some prompt engineering (attached in the ChatGPT chat). I'm considering reading some paper(s) on this topic to see if there's a better entropy function out in the wild. Do you have any recommendations for reads?

Conaclos replied:
Unfortunately I have no expertise in that domain and I didn't take the time to read the literature. Have you tried the new version of the rule on some code bases?

SaadBazaz replied:
I've added more test cases in valid.js, it was able to clear at least 1-2 more false positives (still leaves some leftover ones). Not a HUGE impact though.

I will see if I can find time to read through the literature, however, I might defer it to one of my company's team members if it takes too long.

Conaclos commented on crates/biome_analyze/src/rule.rs:
```diff
@@ -210,7 +213,8 @@ impl RuleSource {
             | Self::EslintBarrelFiles(rule_name)
             | Self::EslintN(rule_name)
-            | Self::EslintNext(rule_name)
+            | Self::Stylelint(rule_name)
+            | Self::EslintNoSecrets(rule_name) => rule_name,
```
```suggestion
            | Self::EslintNext(rule_name)
            | Self::EslintNoSecrets(rule_name)
```
